File: examples/ch13/concept_strategy.py
"""Strategy — canonical conceptual example (runnable).

Encapsulate interchangeable algorithms behind a common interface and inject the
chosen one. The canonical class form is what the name denotes; in Python a
strategy is usually just a function in a registry.
"""

from __future__ import annotations

from abc import ABC, abstractmethod
from collections.abc import Callable


# --- Canonical: a Strategy interface, injected into a Context ----------------

class Strategy(ABC):
    @abstractmethod
    def execute(self, data: list[int]) -> list[int]: ...


class AscendingStrategy(Strategy):
    def execute(self, data: list[int]) -> list[int]:
        return sorted(data)


class DescendingStrategy(Strategy):
    def execute(self, data: list[int]) -> list[int]:
        return sorted(data, reverse=True)


class Context:
    def __init__(self, strategy: Strategy) -> None:
        self._strategy = strategy

    def do_work(self, data: list[int]) -> list[int]:
        return self._strategy.execute(data)   # delegates to the injected strategy


# --- Pythonic: a strategy is a function; a registry selects one --------------

SortStrategy = Callable[[list[int]], list[int]]

STRATEGIES: dict[str, SortStrategy] = {
    "asc": sorted,
    "desc": lambda d: sorted(d, reverse=True),
}


if __name__ == "__main__":
    print(Context(DescendingStrategy()).do_work([3, 1, 2]))  # [3, 2, 1]
    print(STRATEGIES["asc"]([3, 1, 2]))                       # [1, 2, 3]

File: examples/ch13/ts/conceptStrategy.ts
// Strategy — canonical conceptual example. Run: npx tsx conceptStrategy.ts
//
// Encapsulate interchangeable algorithms behind a common interface and inject the
// chosen one. The canonical class form is what the name denotes; in TS a strategy
// is usually just a function in a record.

// --- Canonical: a Strategy interface, injected into a Context ----------------

export interface Strategy {
  execute(data: number[]): number[];
}

export class AscendingStrategy implements Strategy {
  execute(data: number[]): number[] {
    return [...data].sort((a, b) => a - b);
  }
}

export class DescendingStrategy implements Strategy {
  execute(data: number[]): number[] {
    return [...data].sort((a, b) => b - a);
  }
}

export class Context {
  constructor(private readonly strategy: Strategy) {}
  doWork(data: number[]): number[] {
    return this.strategy.execute(data); // delegates to the injected strategy
  }
}

// --- Idiomatic: a strategy is a function; a record selects one --------------

export type SortStrategy = (data: number[]) => number[];

export const STRATEGIES: Record<string, SortStrategy> = {
  asc: (d) => [...d].sort((a, b) => a - b),
  desc: (d) => [...d].sort((a, b) => b - a),
};

// Demo
console.log(new Context(new DescendingStrategy()).doWork([3, 1, 2])); // [3, 2, 1]
console.log(STRATEGIES.asc?.([3, 1, 2])); // [1, 2, 3]
